### PR TITLE
Ifpack2: Chebyshev+Residual schedule

### PR DESCRIPTION
**Pull request by cgcgcg**
@trilinos/ifpack2 

## Motivation
Use SpMV launch parameters and policies for Chebyshev and ScaledDampedResidual.

### CODE REVIEW

**Comment by cgcgcg** [comment]
@jjellio

**Comment by jjellio** [comment]
I am rebuilding mini-em with:
(this should be the same as passing Kokkos::Static)

```
diff --git a/packages/ifpack2/src/Ifpack2_Details_ChebyshevKernel_def.hpp b/packages/ifpack2/src/Ifpack2_Details_ChebyshevKernel_def.hpp
index baa1f207d0..6f0048b0b1 100644
--- a/packages/ifpack2/src/Ifpack2_Details_ChebyshevKernel_def.hpp
+++ b/packages/ifpack2/src/Ifpack2_Details_ChebyshevKernel_def.hpp
@@ -278,7 +278,7 @@ chebyshev_kernel_vector
   using Kokkos::Dynamic;
   using Kokkos::Schedule;
   using Kokkos::TeamPolicy;
-  using policy_type = TeamPolicy<execution_space, Schedule<Dynamic>>;
+  using policy_type = TeamPolicy<execution_space>;
   const char kernel_label[] = "chebyshev_kernel_vector";
   policy_type policy (1, 1);
   if (team_size < 0) {
diff --git a/packages/ifpack2/src/Ifpack2_Details_ScaledDampedResidual_def.hpp b/packages/ifpack2/src/Ifpack2_Details_ScaledDampedResidual_def.hpp
index 68cade49dc..f675d82310 100644
--- a/packages/ifpack2/src/Ifpack2_Details_ScaledDampedResidual_def.hpp
+++ b/packages/ifpack2/src/Ifpack2_Details_ScaledDampedResidual_def.hpp
@@ -265,7 +265,7 @@ scaled_damped_residual_vector
   using Kokkos::Dynamic;
   using Kokkos::Schedule;
   using Kokkos::TeamPolicy;
-  using policy_type = TeamPolicy<execution_space, Schedule<Dynamic>>;
+  using policy_type = TeamPolicy<execution_space>;
   const char kernel_label[] = "scaled_damped_residual_vector";
   policy_type policy (1, 1);
   if (team_size < 0) {
```

File: packages/ifpack2/src/Ifpack2_Details_ChebyshevKernel_def.hpp [hunk 1]
```diff
@@ -332,7 +276,10 @@ chebyshev_kernel_vector
                                      use_beta,
                                      true>;
       functor_type func (alpha, w, d, b, A, x_colMap, x_domMap, beta, rows_per_team);
-      Kokkos::parallel_for (kernel_label, policy, func);
+      if(A.nnz()>10000000)
```
Comment (jhux2) [inline packages/ifpack2/src/Ifpack2_Details_ChebyshevKernel_def.hpp:279]: How was this number chosen -- does it just follow what KokkosKernels does?

File: packages/ifpack2/src/Ifpack2_Details_ChebyshevKernel_def.hpp [hunk 2]
```diff
@@ -342,7 +289,10 @@ chebyshev_kernel_vector
                                      use_beta,
                                      false>;
       functor_type func (alpha, w, d, b, A, x_colMap, x_domMap, beta, rows_per_team);
-      Kokkos::parallel_for (kernel_label, policy, func);
+      if(A.nnz()>10000000)
```
Comment (jhux2) [inline packages/ifpack2/src/Ifpack2_Details_ChebyshevKernel_def.hpp:292]: Same question as above.

File: packages/ifpack2/src/Ifpack2_Details_ChebyshevKernel_def.hpp (hunk 1 above)
Comment (cgcgcg) [reply packages/ifpack2/src/Ifpack2_Details_ChebyshevKernel_def.hpp:279]: Yup. That's the same number as here:
https://github.com/trilinos/Trilinos/blob/442a24ba155050971e7cba30fa2c22fdda9ebce9/packages/kokkos-kernels/src/sparse/impl/KokkosSparse_spmv_impl.hpp#L314-L331

**Comment by jjellio** [comment]
Data off Mutrino KNL, 4x16

dynamic 'yes' : the original build
cg: this PR
others: my patches above 

Threads | dynamic | BelosLOWS
---|---|---
16| yes           | 33.6922
64| yes           | 326.716
16| no-cheby         | 29.9115
64|  no-cheby          | 33.5027
16| no-cheb-and-res| 29.9493
64| no-cheb-and-res| 32.8812
16| cg (all of the above)            | 29.9338
64| cg (all of the above)           | 33.6452

So atleast on KNL, this looks OK.

I'll try it with HSW + HT=2

File: packages/ifpack2/src/Ifpack2_Details_ChebyshevKernel_def.hpp (hunk 1 above)
Comment (jjellio) [reply packages/ifpack2/src/Ifpack2_Details_ChebyshevKernel_def.hpp:279]: There was the work I did a while back, that showed the SpMV struggled based on NNZ **per row** (on Cuda);  Less so than total nnz. Christian Trott tuned that via the vector length.

I wonder if these parameters need to be set based on the average nnz pe row or simply the number of rows.  Raw nnz seems odd (don't all of the kernels operate on rows?).

**Pull request merged**